### PR TITLE
Extend watch/release channel logic in DataNode

### PR DESCRIPTION
**Pull request by congqixia**
Related #15846 

/kind enhancement

Signed-off-by: Congqi Xia <congqi.xia@zilliz.com>

### CODE REVIEW

File: internal/datanode/data_node.go [hunk 1]
```diff
@@ -314,6 +316,13 @@ func (node *DataNode) handleWatchInfo(e *event, key string, data []byte) {
 			return
 		}
 
+		if watchInfo.State != datapb.ChannelWatchState_ToWatch && // start watch
+			watchInfo.State != datapb.ChannelWatchState_ToRelease && // start release
+			watchInfo.State != datapb.ChannelWatchState_Uncomplete { // legacy state, equal to ToWatch
+			log.Warn("watch channel state in result state", zap.String("state", watchInfo.State.String()))
```
Comment (soothing-rain) [inline internal/datanode/data_node.go:322]: How about something like "Data node is handling a PUT event, while the channel watch info was in wrong state"

File: internal/datanode/data_node.go [hunk 2]
```diff
@@ -314,6 +316,13 @@ func (node *DataNode) handleWatchInfo(e *event, key string, data []byte) {
 			return
 		}
 
+		if watchInfo.State != datapb.ChannelWatchState_ToWatch && // start watch
+			watchInfo.State != datapb.ChannelWatchState_ToRelease && // start release
+			watchInfo.State != datapb.ChannelWatchState_Uncomplete { // legacy state, equal to ToWatch
```
Comment (soothing-rain) [inline internal/datanode/data_node.go:321]: I can see the same code block appeared here and again below  (event_manager.go:118). Maybe worth making a function for them?

File: internal/datanode/data_node.go [hunk 3]
```diff
@@ -350,9 +356,9 @@ func parsePutEventData(data []byte) (*datapb.ChannelWatchInfo, error) {
 		return nil, fmt.Errorf("invalid event data: fail to parse ChannelWatchInfo, err: %v", err)
 	}
 
-	if watchInfo.State == datapb.ChannelWatchState_Complete {
-		return nil, fmt.Errorf("invalid event: event state is already ChannelWatchState_Compele")
-	}
+	//if watchInfo.State == datapb.ChannelWatchState_Complete {
```
Comment (soothing-rain) [inline internal/datanode/data_node.go:359]: Are these lines commented out on purpose?

File: internal/datanode/data_node.go [hunk 4]
```diff
@@ -367,28 +373,51 @@ func parseDeleteEventKey(key string) string {
...
+			watchInfo.State = datapb.ChannelWatchState_ReleaseSuccess
+		}
 	}
-	log.Info("handle put event: new data sync service success", zap.String("vChanName", vChanName))
 
-	watchInfo.State = datapb.ChannelWatchState_Complete
 	v, err := proto.Marshal(watchInfo)
 	if err != nil {
-		return fmt.Errorf("fail to marshal watchInfo with complete state, vChanName: %s, err: %v", vChanName, err)
+		return fmt.Errorf("fail to marshal watchInfo with state, vChanName: %s, err: %v", vChanName, err)
```
Comment (soothing-rain) [inline internal/datanode/data_node.go:412]: nit: Shall we add WatchInfo state back in the error message?

File: internal/datanode/data_node.go [hunk 5]
```diff
@@ -367,28 +373,51 @@ func parseDeleteEventKey(key string) string {
...
-	log.Info("handle put event: try to save completed state", zap.String("key", k))
+	log.Info("handle put event: try to save result state", zap.String("key", k), zap.String("state", watchInfo.State.String()))
 
-	err = node.watchKv.Save(k, string(v))
-	// TODO DataNode unable to save into etcd, may need to panic
+	err = node.watchKv.CompareVersionAndSwap(k, version, string(v))
 	if err != nil {
-		node.releaseFlowgraph(vChanName)
-		return fmt.Errorf("fail to update completed state to etcd, vChanName: %s, err: %v", vChanName, err)
+		return fmt.Errorf("fail to update watch state to etcd, vChanName: %s, err: %w", vChanName, err)
```
Comment (soothing-rain) [inline internal/datanode/data_node.go:420]: nit: Shall we add watch state back in the error message?

File: internal/datanode/event_manager.go [hunk 6]
```diff
@@ -92,10 +78,76 @@ func (e *channelEventManager) Run() {
 	}()
 }
 
+func (e *channelEventManager) retryHandlePutEvent(event event) {
+	deadline := time.Unix(0, event.info.TimeoutTs)
+	countdown := time.Until(deadline)
+	if countdown < 0 {
+		log.Warn("event already timeout", zap.String("vChanName", event.vChanName))
```
Comment (soothing-rain) [inline internal/datanode/event_manager.go:85]: nit: event already timed out

File: internal/datanode/event_manager.go [hunk 7]
```diff
@@ -92,10 +78,76 @@ func (e *channelEventManager) Run() {
 	}()
 }
 
+func (e *channelEventManager) retryHandlePutEvent(event event) {
+	deadline := time.Unix(0, event.info.TimeoutTs)
```
Comment (soothing-rain) [inline internal/datanode/event_manager.go:82]: countdown := time.Until(time.Unix(0, event.info.TimeoutTs)

File: internal/datanode/event_manager.go [hunk 8]
```diff
@@ -92,10 +78,76 @@ func (e *channelEventManager) Run() {
...
+
+			// reset the ticker
+			ticker.Reset(e.retryInterval)
+
+			select {
+			case <-ticker.C:
+				// ticker notify, do another retry
+			case <-timer.C:
+				// timeout
+				log.Warn("event process timeout", zap.String("vChanName", event.vChanName))
```
Comment (soothing-rain) [inline internal/datanode/event_manager.go:108]: event process timed out

File: internal/datanode/data_node.go [hunk 9]
```diff
@@ -367,28 +373,51 @@ func parseDeleteEventKey(key string) string {
...
 }
 
-func (node *DataNode) handlePutEvent(watchInfo *datapb.ChannelWatchInfo) error {
+func (node *DataNode) handlePutEvent(watchInfo *datapb.ChannelWatchInfo, version int64) (err error) {
 	vChanName := watchInfo.GetVchan().GetChannelName()
 
-	if err := node.flowgraphManager.addAndStart(node, watchInfo.GetVchan()); err != nil {
-		return fmt.Errorf("fail to add and start flowgraph for vChanName: %s, err: %v", vChanName, err)
+	switch watchInfo.State {
+	case datapb.ChannelWatchState_Uncomplete, datapb.ChannelWatchState_ToWatch:
```
Comment (soothing-rain) [inline internal/datanode/data_node.go:380]: nit: Could use protobuf alias: `option allow_alias = true`(https://developers.google.com/protocol-buffers/docs/proto3)

Comment (congqixia) [reply internal/datanode/data_node.go:374]: Since we need to deprecated old state, maybe its better to separate `Complete` and `WatchSuccess`.

**Pull request merged**